src/buildstream_plugins/sources/cargo.py

```diff
@@ -294,6 +310,268 @@ def _get_mirror_file(self, sha=None):
+            json.dump({"files": {}, "package": None}, f)
+
+        shutil.rmtree(tmp_dir, ignore_errors=True)
+
+    # is_cached()
+    #
+    # Get whether we have a local cached version of the git commit
+    #
+    # Returns:
+    #   (bool): Whether we are cached or not
+    #
+    def is_cached(self):
+        with Repo(self._get_mirror_dir(), bare=True) as repo:
+            return self.commit.encode() in repo
+
+    # is_resolved()
+    #
+    # Get whether the current git repo is resolved
+    #
+    # Returns:
+    #   (bool): Always true since we always have a commit
+    #
+    def is_resolved(self):
+        return True
+
+    ########################################################
+    #                   Private helpers                    #
+    ########################################################
+
+    # _get_mirror_dir()
+    #
+    # Gets the local mirror directory for this upstream git repository
+    #
+    def _get_mirror_dir(self):
+        if self.repo.endswith(".git"):
+            norm_url = self.repo[:-4]
+        else:
+            norm_url = self.repo
+
+        return os.path.join(
+            self.cargo.get_mirror_directory(),
+            utils.url_directory_name(norm_url) + ".git",
+        )
+
+    # _get_url()
+    #
+    # Gets the git URL to download this crate from
+    #
+    # Args:
+    #    alias (str|None): The URL alias to apply, if any
+    #
+    # Returns:
+    #    (str): The URL for this crate
+    #
+    def _get_url(self, alias=None):
+        return self.cargo.translate_url(self.cargo.repo, alias_override=alias)
+
+    # _mirror_repo()
+    #
+    # Returns the mirror repo, initialized if it doesn not exist yet
+    #
+    # Returns:
+    #    (Repo): The mirror repo crate
+    #
+    def _mirror_repo(self):
+        try:
+            return Repo.init_bare(self._get_mirror_dir(), mkdir=True)
+        except FileExistsError:
+            return Repo(self._get_mirror_dir(), bare=True)
+
+
+# config_inherit_workspace()
+#
+# Adds inherited values to config
+#
+# Args:
+#    config (dict): Crate config
+#    workspace_config (dict): Workspace config
+def config_inherit_workspace(config, workspace_config):
+    workspace_deps = workspace_config.get("dependencies")
+    if workspace_deps is not None:
+        dependencies = []
+        for key in ["dependencies", "dev-dependencies", "build-dependencies"]:
+            if key in config:
+                dependencies.append(config[key])
+
+        if "traget" in config:
+            for target in config["target"].values():
+                if "dependencies" in target:
+                    dependencies.append(target["dependencies"])
+
+        for deps in dependencies:
+            inherit_deps(deps, workspace_deps)
+
+    workspace_package = workspace_config.get("package")
+    if workspace_package is not None:
+        inherit_package(config["package"], workspace_package)
+
+
+# inherit_package()
+#
+# Adds inherited [package] entries to config
```

Comment:
Function docs for this function, and similar ones (`config_inherit_workspace`, `inherit_deps`, etc), aren't very helpful.

> Adds inherited [package] entries to config

Inherited from where? Adds to what config?

Reply:
Maybe there's some documentation of the Cargo.toml file that you can reference in these function comments, so it's clear what exactly you're doing and why? Right now the code is just twiddling some entries in some dictionaries, and unless I go looking for docs I have no way to know whether it's correct (which by the way I'm not doing for this review; I'll assume that you're manipulating the toml files correctly :smile:)